File: cloud-functions/internal/handler/mcp.go
// Package handler — Standard MCP Streamable HTTP endpoint (/mcp).
//
// Implements MCP JSON-RPC 2.0 over HTTP for external AI clients
// (Cursor, Claude Desktop, VS Code / Cline, Cherry Studio, …).
//
// Supported methods:
//
//	initialize              — handshake; returns server info and capabilities
//	ping                    — heartbeat
//	tools/list              — list available tools
//	tools/call              — execute a tool
//	notifications/initialized — notification (no response needed)
//
// GET  /mcp — service discovery (cached 60 s)
// POST /mcp — main JSON-RPC endpoint; responds with JSON or SSE depending on Accept header
// DELETE /mcp — session termination (no-op)
package handler

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"

	"cloud-functions/internal/config"
	"cloud-functions/internal/knowledge"
)

// ── constants ─────────────────────────────────────────────────────────────────

const (
	mcpServerName      = "vector-mcp-edge-knowledge-mcp"
	mcpServerVersion   = "2.0.0"
	mcpProtocolVersion = "2025-03-26"
)

// ── embed static text content ─────────────────────────────────────────────────

//go:embed static/project_info.md
var mcpProjectInfoText string

//go:embed static/quickstart.md
var mcpQuickstartText string

//go:embed static/solutions.md
var mcpSolutionsText string

// ── MCP tool definitions (tools/list response) ────────────────────────────────

var mcpTools = []map[string]interface{}{
	{
		"name": "query_knowledge_base",
		"description": "Search the VitePress MCP documentation knowledge base using semantic vector search. " +
			"Use this to find information about MCP server setup on EdgeOne Pages, CNB knowledge base " +
			"configuration, VitePress customisation, and deployment workflows.",
		"inputSchema": map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"query": map[string]interface{}{
					"type":        "string",
					"description": "Search query in natural language (Chinese or English). E.g. '如何配置 MCP Server'",
				},
				"keyword": map[string]interface{}{
					"type":        "string",
					"description": "Optional keywords separated by semicolons. E.g. 'EdgeOne;MCP;部署'",
				},
				"top_k": map[string]interface{}{
					"type":        "number",
					"description": "Maximum results to return (default 5, range 1–10).",
				},
			},
			"required": []string{"query"},
		},
	},
	{
		"name":        "get_project_info",
		"description": "Get an overview of the VitePress MCP project: purpose, tech stack, features, and links.",
		"inputSchema": map[string]interface{}{"type": "object", "properties": map[string]interface{}{}},
	},
	{
		"name":        "get_quickstart",
		"description": "Get the step-by-step quick-start guide for setting up VitePress MCP smart search.",
		"inputSchema": map[string]interface{}{"type": "object", "properties": map[string]interface{}{}},
	},
	{
		"name":        "get_solutions",
		"description": "Compare the available solutions: Go Cloud Function (recommended) vs legacy JS MCP and Go RAG.",
		"inputSchema": map[string]interface{}{"type": "object", "properties": map[string]interface{}{}},
	},
}

// ── JSON-RPC types ────────────────────────────────────────────────────────────

type mcpRequest struct {
	JSONRPC string                 `json:"jsonrpc"`
	ID      interface{}            `json:"id"`
	Method  string                 `json:"method"`
	Params  map[string]interface{} `json:"params"`
}

// ── JSON-RPC response helpers ─────────────────────────────────────────────────

func mcpOK(id interface{}, result interface{}) map[string]interface{} {
	return map[string]interface{}{"jsonrpc": "2.0", "id": id, "result": result}
}

func mcpError(id interface{}, code int, message string) map[string]interface{} {
	return map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      id,
		"error":   map[string]interface{}{"code": code, "message": message},
	}
}

func mcpText(text string) map[string]interface{} {
	return map[string]interface{}{
		"content": []map[string]interface{}{{"type": "text", "text": text}},
	}
}

func mcpErrorContent(text string) map[string]interface{} {
	return map[string]interface{}{
		"content": []map[string]interface{}{{"type": "text", "text": text}},
		"isError": true,
	}
}

// ── protocol method handlers ──────────────────────────────────────────────────

func mcpInitialize(req mcpRequest) map[string]interface{} {
	return mcpOK(req.ID, map[string]interface{}{
		"protocolVersion": mcpProtocolVersion,
		"capabilities":    map[string]interface{}{"tools": map[string]interface{}{"listChanged": false}},
		"serverInfo":      map[string]interface{}{"name": mcpServerName, "version": mcpServerVersion},
	})
}

func mcpPing(req mcpRequest) map[string]interface{} {
	return mcpOK(req.ID, map[string]interface{}{})
}

func mcpToolsList(req mcpRequest) map[string]interface{} {
	return mcpOK(req.ID, map[string]interface{}{"tools": mcpTools})
}

func mcpToolsCall(req mcpRequest, cfg config.Config) map[string]interface{} {
	name, _ := req.Params["name"].(string)
	args, _ := req.Params["arguments"].(map[string]interface{})

	switch name {
	case "query_knowledge_base":
		return mcpQueryKnowledge(req.ID, args, cfg)
	case "get_project_info":
		return mcpOK(req.ID, mcpText(mcpProjectInfoText))
	case "get_quickstart":
		return mcpOK(req.ID, mcpText(mcpQuickstartText))
	case "get_solutions":
		return mcpOK(req.ID, mcpText(mcpSolutionsText))
	default:
		return mcpError(req.ID, -32602, fmt.Sprintf("Unknown tool: %s", name))
	}
}

func mcpQueryKnowledge(id interface{}, args map[string]interface{}, cfg config.Config) map[string]interface{} {
	if args == nil {
		return mcpError(id, -32602, "Missing required parameter: query")
	}
	query, _ := args["query"].(string)
	if query == "" {
		return mcpError(id, -32602, "Missing required parameter: query")
	}
	kw, _ := args["keyword"].(string)
	topK := 5
	if v, ok := args["top_k"].(float64); ok {
		topK = int(math.Max(1, math.Min(10, math.Floor(v))))
	}

	result, err := knowledge.Query(cfg, query, kw, topK)
	if err != nil {
		return mcpOK(id, mcpErrorContent(fmt.Sprintf("Error querying knowledge base: %v", err)))
	}
	content := knowledge.FormatResult(result)
	if content == "" {
		content = "No results found for the given query."
	}
	return mcpOK(id, mcpText(content))
}

// mcpDispatch routes a single JSON-RPC request. Returns nil for notifications.
func mcpDispatch(req mcpRequest, cfg config.Config) map[string]interface{} {
	switch req.Method {
	case "initialize":
		return mcpInitialize(req)
	case "ping":
		return mcpPing(req)
	case "tools/list":
		return mcpToolsList(req)
	case "tools/call":
		return mcpToolsCall(req, cfg)
	case "notifications/initialized":
		return nil // notification — no response
	default:
		if req.ID != nil {
			fmt.Printf("[MCP] unknown method: %s\n", req.Method)
		}
		return mcpError(req.ID, -32601, fmt.Sprintf("Method not found: %s", req.Method))
	}
}

// ── HTTP entry point ──────────────────────────────────────────────────────────

// MCPStreamable returns the Gin handler for the standard MCP Streamable HTTP transport.
// Register with r.Any("/mcp", handler.MCPStreamable(cfg)).
func MCPStreamable(cfg config.Config) gin.HandlerFunc {
	return func(c *gin.Context) {
		switch c.Request.Method {

		case http.MethodGet:
			// Service-discovery endpoint; 60 s cache to reduce heartbeat overhead.
			c.Header("Cache-Control", "public, max-age=60")
			c.JSON(http.StatusOK, gin.H{
				"name":            mcpServerName,
				"version":         mcpServerVersion,
				"description":     "MCP server for VitePress MCP 智能检索. Use POST to interact.",
				"protocolVersion": mcpProtocolVersion,
				"availableTools":  mcpToolSummary(),
			})

		case http.MethodDelete:
			c.Status(http.StatusOK)

		case http.MethodPost:
			if !strings.Contains(c.GetHeader("Content-Type"), "application/json") {
				c.JSON(http.StatusUnsupportedMediaType,
					mcpError(nil, -32700, "Content-Type must be application/json"))
				return
			}
			body, err := io.ReadAll(c.Request.Body)
			if err != nil {
				c.JSON(http.StatusBadRequest, mcpError(nil, -32700, "Failed to read request body"))
				return
			}

			useSSE := strings.Contains(c.GetHeader("Accept"), "text/event-stream")
			trimmed := bytes.TrimSpace(body)

			if len(trimmed) > 0 && trimmed[0] == '[' {
				// Batch request
				var batch []mcpRequest
				if err := json.Unmarshal(body, &batch); err != nil {
					c.JSON(http.StatusBadRequest, mcpError(nil, -32700, "Parse error: Invalid JSON"))
					return
				}
				var results []map[string]interface{}
				for _, req := range batch {
					if r := mcpDispatch(req, cfg); r != nil {
						results = append(results, r)
					}
				}
				if len(results) == 0 {
					c.Status(http.StatusAccepted)
					return
				}
				if useSSE {
					writeMCPSSEBatch(c, results)
				} else {
					c.JSON(http.StatusOK, results)
				}
			} else {
				// Single request
				var req mcpRequest
				if err := json.Unmarshal(body, &req); err != nil {
					c.JSON(http.StatusBadRequest, mcpError(nil, -32700, "Parse error: Invalid JSON"))
					return
				}
				result := mcpDispatch(req, cfg)
				if result == nil {
					c.Status(http.StatusAccepted)
					return
				}
				if useSSE {
					writeMCPSSE(c, result)
				} else {
					c.JSON(http.StatusOK, result)
				}
			}

		default:
			c.JSON(http.StatusMethodNotAllowed, mcpError(nil, -32600, "Method not allowed"))
		}
	}
}

// ── SSE helpers ───────────────────────────────────────────────────────────────

func mcpToolSummary() []map[string]interface{} {
	out := make([]map[string]interface{}, 0, len(mcpTools))
	for _, t := range mcpTools {
		out = append(out, map[string]interface{}{
			"name":        t["name"],
			"description": t["description"],
		})
	}
	return out
}

func writeMCPSSE(c *gin.Context, data map[string]interface{}) {
	c.Header("Content-Type", "text/event-stream")
	c.Header("Cache-Control", "no-cache")
	c.Header("Connection", "keep-alive")
	b, _ := json.Marshal(data)
	fmt.Fprintf(c.Writer, "event: message\ndata: %s\n\n", string(b))
	c.Writer.Flush()
}

func writeMCPSSEBatch(c *gin.Context, data []map[string]interface{}) {
	c.Header("Content-Type", "text/event-stream")
	c.Header("Cache-Control", "no-cache")
	c.Header("Connection", "keep-alive")
	for _, item := range data {
		b, _ := json.Marshal(item)
		fmt.Fprintf(c.Writer, "event: message\ndata: %s\n\n", string(b))
	}
	c.Writer.Flush()
}

File: cloud-functions/mcp.js
const SERVER_NAME = 'vector-mcp-edge-knowledge-mcp'
const SERVER_VERSION = '3.0.0'
const PROTOCOL_VERSION = '2025-03-26'

const PROJECT_INFO = `# Vector MCP Edge

Vector MCP Edge is a tutorial and runnable demo for adding semantic documentation search to a VitePress site.

- CNB knowledge base chunks and vectorizes Markdown documents.
- EdgeOne Makers Node Cloud Function exposes the standard /mcp endpoint.
- EdgeOne Makers hosted Agent provides /chat conversations and /stop cancellation.
- Makers AI Gateway and Store provide model access and persistent sessions.

Repository: https://cnb.cool/shenzhen/lecturer/vector-mcp-edge
Documentation: https://vector-mcp-edge.mintimate.cn/`

const QUICKSTART = `# Quick Start

1. Host the Markdown repository on CNB and enable the knowledge-base pipeline.
2. Declare AI_GATEWAY_* and CNB_KNOWLEDGE_BASE_* in .env.example.
3. Configure the CNB URL and token with edgeone makers env set.
4. Run PAGES_SOURCE=skills edgeone makers dev.
5. Deploy with PAGES_SOURCE=skills edgeone makers deploy.
6. Connect external AI clients to https://<domain>/mcp.
7. Call POST /chat with a makers-conversation-id header.`

const SOLUTIONS = `# Solution Overview

The current EdgeOne Makers architecture uses a path-level Node Cloud Function for MCP and a hosted Agent for model inference, knowledge retrieval, SSE, sessions, and cancellation. Both share the same CNB knowledge base.

Archived approaches include JS-only MCP, self-hosted Go RAG, and a catch-all Go Function with browser-orchestrated Tool Calling.`

const tools = [
  {
    name: 'query_knowledge_base',
    description: 'Search the VitePress MCP documentation knowledge base using semantic vector search. Use this for EdgeOne Makers, hosted Agents, CNB knowledge bases, VitePress, MCP, configuration, and deployment workflows.',
    inputSchema: {
      type: 'object',
      properties: {
        query: { type: 'string', description: 'Natural-language search query in Chinese or English.' },
        keyword: { type: 'string', description: 'Optional semicolon-separated keywords.' },
        top_k: { type: 'number', description: 'Maximum results (default 5, range 1-10).' }
      },
      required: ['query']
    }
  },
  {
    name: 'get_project_info',
    description: 'Get the current Vector MCP Edge architecture, purpose, and links.',
    inputSchema: { type: 'object', properties: {} }
  },
  {
    name: 'get_quickstart',
    description: 'Get the current EdgeOne Makers quick-start guide.',
    inputSchema: { type: 'object', properties: {} }
  },
  {
    name: 'get_solutions',
    description: 'Compare the current Makers MCP + hosted Agent architecture with legacy solutions.',
    inputSchema: { type: 'object', properties: {} }
  }
]

export async function onRequest(context) {
  const request = context.request
  if (request.method === 'OPTIONS') return withCors(new Response(null, { status: 204 }))
  if (request.method === 'GET') {
    return json({
      name: SERVER_NAME,
      version: SERVER_VERSION,
      description: 'MCP server for VitePress MCP smart search. Use POST to interact.',
      protocolVersion: PROTOCOL_VERSION,
      availableTools: tools.map(({ name, description }) => ({ name, description }))
    }, 200, { 'Cache-Control': 'public, max-age=60' })
  }
  if (request.method === 'DELETE') return withCors(new Response(null, { status: 200 }))
  if (request.method !== 'POST') return json(rpcError(null, -32600, 'Method not allowed'), 405)

  if (!request.headers.get('content-type')?.includes('application/json')) {
    return json(rpcError(null, -32700, 'Content-Type must be application/json'), 415)
  }

  let payload
  try {
    payload = await request.json()
  } catch {
    return json(rpcError(null, -32700, 'Parse error: Invalid JSON'), 400)
  }

  const useSSE = request.headers.get('accept')?.includes('text/event-stream')
  if (Array.isArray(payload)) {
    const results = (await Promise.all(payload.map(item => dispatch(item, context.env)))).filter(Boolean)
    if (!results.length) return withCors(new Response(null, { status: 202 }))
    return useSSE ? sse(results) : json(results)
  }

  const result = await dispatch(payload, context.env)
  if (!result) return withCors(new Response(null, { status: 202 }))
  return useSSE ? sse([result]) : json(result)
}

async function dispatch(request, env) {
  switch (request?.method) {
    case 'initialize':
      return rpcOK(request.id, {
        protocolVersion: PROTOCOL_VERSION,
        capabilities: { tools: { listChanged: false } },
        serverInfo: { name: SERVER_NAME, version: SERVER_VERSION }
      })
    case 'ping':
      return rpcOK(request.id, {})
    case 'tools/list':
      return rpcOK(request.id, { tools })
    case 'tools/call':
      return callTool(request.id, request.params, env)
    case 'notifications/initialized':
      return null
    default:
      return rpcError(request?.id ?? null, -32601, `Method not found: ${request?.method ?? ''}`)
  }
}

async function callTool(id, params, env) {
  const name = params?.name
  const args = params?.arguments ?? {}
  if (name === 'get_project_info') return rpcOK(id, textContent(PROJECT_INFO))
  if (name === 'get_quickstart') return rpcOK(id, textContent(QUICKSTART))
  if (name === 'get_solutions') return rpcOK(id, textContent(SOLUTIONS))
  if (name !== 'query_knowledge_base') return rpcError(id, -32602, `Unknown tool: ${name}`)
  if (typeof args.query !== 'string' || !args.query.trim()) {
    return rpcError(id, -32602, 'Missing required parameter: query')
  }

  try {
    const content = await queryKnowledgeBase(args, env)
    return rpcOK(id, textContent(content || 'No results found for the given query.'))
  } catch (error) {
    return rpcOK(id, { ...textContent(`Error querying knowledge base: ${error.message}`), isError: true })
  }
}

async function queryKnowledgeBase(args, env) {
  const url = env.CNB_KNOWLEDGE_BASE_URL || env.KNOWLEDGE_API_URL
  const token = env.CNB_KNOWLEDGE_BASE_TOKEN || env.CNB_TOKEN
  if (!url || !token) throw new Error('CNB knowledge-base environment variables are missing')

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: token,
      'Content-Type': 'application/json; charset=utf-8'
    },
    body: JSON.stringify({
      query: args.query,
      ...(args.keyword ? { keyword: args.keyword } : {}),
      top_k: clampTopK(args.top_k)
    })
  })
  if (!response.ok) throw new Error(`knowledge API ${response.status}: ${await response.text()}`)
  return formatKnowledgePayload(await response.json())
}

function formatKnowledgePayload(payload) {
  const items = collectItems(payload)
  if (!items.length) return JSON.stringify(payload, null, 2)
  return items.map((item, index) => {
    const metadata = isRecord(item.metadata) ? item.metadata : {}
    const title = firstString(metadata, ['name', 'title', 'path']) || 'Untitled'
    const content = firstString(item, ['chunk', 'content', 'text', 'pageContent', 'page_content']) || JSON.stringify(item)
    const score = typeof item.score === 'number' ? ` (relevance: ${item.score.toFixed(3)})` : ''
    const source = firstString(metadata, ['url', 'source'])
    return `## ${index + 1}. ${title}${score}\n\n${content}${source ? `\nSource: ${source}` : ''}`
  }).join('\n\n---\n\n')
}

function collectItems(value) {
  if (Array.isArray(value)) return value
  if (!isRecord(value)) return []
  for (const key of ['data', 'results', 'documents', 'chunks', 'items', 'list']) {
    if (Array.isArray(value[key])) return value[key]
    const nested = collectItems(value[key])
    if (nested.length) return nested
  }
  return [value]
}

function clampTopK(value) {
  const number = Number(value)
  return Number.isFinite(number) ? Math.max(1, Math.min(10, Math.floor(number))) : 5
}

function firstString(value, keys) {
  for (const key of keys) if (typeof value?.[key] === 'string' && value[key]) return value[key]
  return ''
}

function isRecord(value) {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}

function rpcOK(id, result) {
  return { jsonrpc: '2.0', id, result }
}

function rpcError(id, code, message) {
  return { jsonrpc: '2.0', id, error: { code, message } }
}

function textContent(text) {
  return { content: [{ type: 'text', text }] }
}

function json(body, status = 200, extraHeaders = {}) {
  return withCors(Response.json(body, { status, headers: extraHeaders }))
}

function sse(items) {
  const body = items.map(item => `event: message\ndata: ${JSON.stringify(item)}\n\n`).join('')
  return withCors(new Response(body, {
    headers: {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      Connection: 'keep-alive',
      'X-Accel-Buffering': 'no'
    }
  }))
}

function withCors(response) {
  const headers = new Headers(response.headers)
  headers.set('Access-Control-Allow-Origin', '*')
  headers.set('Access-Control-Allow-Methods', 'GET, POST, DELETE, OPTIONS')
  headers.set('Access-Control-Allow-Headers', 'Content-Type, Accept, Authorization, Mcp-Session-Id')
  headers.set('Access-Control-Expose-Headers', 'Mcp-Session-Id')
  return new Response(response.body, { status: response.status, headers })
}
